packages/react-stylesheets/src/StylesheetContext.tsx

```diff
@@ -0,0 +1,61 @@
+import * as React from 'react';
+export type HTMLDirection = 'rtl' | 'ltr' | 'auto';
+
+export interface StylesheetContextType {
+  target: Document | undefined;
+  register: (stylesheets: string[], context: StylesheetContextType) => void;
+  styleCache: WeakMap<Document, Map<string, boolean>>;
+  enqueuedSheets: string[];
+  renderSheets: (stylesheets: string[], context: StylesheetContextType) => void;
+}
+
+/**
+ * Default registration function for stylesheets.
+ */
+export const register = (stylesheets: string[], context: StylesheetContextType) => {
+  const { styleCache, target } = context;
+
+  // Grab the style cache for the target document.
+  let targetStylesheets = styleCache.get(target!);
+
+  if (!targetStylesheets) {
+    targetStylesheets = new Map();
+    styleCache.set(target!, targetStylesheets);
+  }
+
+  for (const stylesheet of stylesheets) {
+    if (!targetStylesheets.has(stylesheet)) {
+      context.enqueuedSheets.unshift(stylesheet);
```

Comment:
reminder on Levi's point earlier today - we can't guarantee the correct order based on rendering order?

Reply:
No we can't guarantee, but the same applies to the typical style injection on module import. All it takes is an async import to change the order.

What we can do is avoid flipping order completely in the normal case. Typically an async import is not happening on a sub component but rather on an experience boundary.

Reply:
I am going to think about if this is really necessary. Right now I think it is. But maybe we just need to attack the problem at the styling rule layer.